WIP: Enable firewalld and open necessary ports

oomichi opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This adds a feature which enables firewalld service and open necessary TCP/UDP ports according to [1].

[1]: https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#check-required-ports

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

oomichi commented on contrib/os-services/roles/prepare/tasks/main.yml:
```diff
@@ -21,3 +21,38 @@
+      with_items: "{{ public_port_list_on_node }}"
+      when: inventory_hostname in groups['kube_node']
+
+    - name: Allow ports on libvirt chain of firewalld
```
On clean CentOS7 environment, libvirt chain doesn't exist..
```
$ sudo firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: enp0s3
  sources: 
  services: dhcpv6-client ssh
  ports: 
  protocols: 
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
```

oomichi commented:
/retitle Enable firewalld and open necessary ports

champtar commented:
Hi @oomichi, If you just open to the whole world you don't gain much security. I have some local modifications to get node IP, create ipsets, ...
As there is no official support for firewalld + k8s and reloading firewalld in CentOS 7 == broken node, I've switched to using Calico (and found some fun bug/limitations).
I can push the 2 approach / commit if you want to test  ​them, not sure I'll have time to rework them.

oomichi commented:
> 
> 
> Hi @oomichi, If you just open to the whole world you don't gain much security. I have some local modifications to get node IP, create ipsets, ...
> As there is no official support for firewalld + k8s and reloading firewalld in CentOS 7 == broken node, I've switched to using Calico (and found some fun bug/limitations).
> I can push the 2 approach / commit if you want to test ​them, not sure I'll have time to rework them.

Thank you @champtar 
It is happy to see another approach. That is the reason why I contribute this code to get more feedback :-)

Recently my colleague asked me to enable firewalld on the company work, and this works fine on that.
(After deploying Kubernetes cluster, I needed to open all TCP/UDP ports on libvirt chain for our applications. That can be another work)

The network plugin is matter on this firewall thing as you also mentioned.
We used Weave when disabling firewalld, that worked fine.
However Weave official page said `Weave Net does not work on hosts running iptables 1.8 or above, only with 1.6.` as https://www.weave.works/docs/net/latest/kubernetes/kube-addon/#-things-to-watch-out-for 
then we have switched to use Calico + VXLAN because we need to run the cluster on MS Azure vms also.
(The default Calico doesn't work on MS Azure vms)

oomichi commented:
Tested on clean CentOS7 environment:
```
$ ansible-playbook -e "{enable_service_firewall: true}" -i inventory/sample/hosts.yaml --become --become-user=root --ask-become-pass contrib/os-services/os-services.yml 
BECOME password: 

PLAY [all] **********************************************************************************************************************************************************************************************************************************
Friday 27 August 2021  20:58:10 -0400 (0:00:00.020)       0:00:00.020 ********* 

TASK [Gathering Facts] **********************************************************************************************************************************************************************************************************************
ok: [node1]
Friday 27 August 2021  20:58:12 -0400 (0:00:01.887)       0:00:01.907 ********* 
Friday 27 August 2021  20:58:12 -0400 (0:00:00.052)       0:00:01.960 ********* 
Friday 27 August 2021  20:58:12 -0400 (0:00:00.043)       0:00:02.003 ********* 
Friday 27 August 2021  20:58:12 -0400 (0:00:00.041)       0:00:02.045 ********* 

TASK [prepare : List services] **************************************************************************************************************************************************************************************************************
ok: [node1]
Friday 27 August 2021  20:58:13 -0400 (0:00:00.818)       0:00:02.863 ********* 

TASK [prepare : Enable service firewalld] ***************************************************************************************************************************************************************************************************
ok: [node1]
Friday 27 August 2021  20:58:13 -0400 (0:00:00.482)       0:00:03.346 ********* 

TASK [prepare : Allow ports on public chain of firewalld] ***********************************************************************************************************************************************************************************
changed: [node1] => (item=6443/tcp)
changed: [node1] => (item=2379-2380/tcp)
changed: [node1] => (item=10250/tcp)
changed: [node1] => (item=10251/tcp)
changed: [node1] => (item=10252/tcp)
Friday 27 August 2021  20:58:15 -0400 (0:00:02.089)       0:00:05.435 ********* 

TASK [prepare : Allow ports on public chain of firewalld] ***********************************************************************************************************************************************************************************
ok: [node1] => (item=10250/tcp)
changed: [node1] => (item=30000-32767/tcp)
Friday 27 August 2021  20:58:16 -0400 (0:00:00.804)       0:00:06.240 ********* 

RUNNING HANDLER [prepare : Reload service firewalld] ****************************************************************************************************************************************************************************************
changed: [node1]

PLAY RECAP **********************************************************************************************************************************************************************************************************************************
node1                      : ok=6    changed=3    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0   

Friday 27 August 2021  20:58:18 -0400 (0:00:01.441)       0:00:07.681 ********* 
=============================================================================== 
prepare : Allow ports on public chain of firewalld ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2.09s
Gathering Facts ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.89s
prepare : Reload service firewalld --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 1.44s
prepare : List services -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.82s
prepare : Allow ports on public chain of firewalld ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.80s
prepare : Enable service firewalld --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.48s
prepare : List services -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.05s
prepare : Disable service firewalld -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.04s
prepare : Disable service ufw -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 0.04s
$ sudo firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: enp0s3
  sources: 
  services: dhcpv6-client ssh
  ports: 6443/tcp 2379-2380/tcp 10250/tcp 10251/tcp 10252/tcp 30000-32767/tcp
  protocols: 
  masquerade: no
  forward-ports: 
  source-ports: 
  icmp-blocks: 
  rich rules: 
```

champtar commented:
@oomichi
Calico as host firewall: https://github.com/champtar/kubespray/commit/633a6f8e9d84b7282c97c72e36b74773c028f0e9
Firewalld as host firewall: https://github.com/champtar/kubespray/commit/b95a8e4f0690dd175af02d0416fe8412e3f58d6f

oomichi commented:
> 
> 
> @oomichi
> Calico as host firewall: [champtar@633a6f8](https://github.com/champtar/kubespray/commit/633a6f8e9d84b7282c97c72e36b74773c028f0e9)
> Firewalld as host firewall: [champtar@b95a8e4](https://github.com/champtar/kubespray/commit/b95a8e4f0690dd175af02d0416fe8412e3f58d6f)

Thank you so much for sharing it @champtar  :-)
Let me try it.

oomichi commented:
> 
> 
> > @oomichi
> > Calico as host firewall: [champtar@633a6f8](https://github.com/champtar/kubespray/commit/633a6f8e9d84b7282c97c72e36b74773c028f0e9)
> > Firewalld as host firewall: [champtar@b95a8e4](https://github.com/champtar/kubespray/commit/b95a8e4f0690dd175af02d0416fe8412e3f58d6f)
> 
> Thank you so much for sharing it @champtar :-)
> Let me try it.

I tested 2 above commits by applying into my local environment, and that doesn't work for me unfortunately.
@champtar Are you using Ubuntu or CentOS or other operating system?

champtar commented:
CentOS 7 and 8

champtar commented:
Both commits at the same time ?

oomichi commented:
> 
> 
> Both commits at the same time ?

Yes, applied both commits to Kubespray 2.16 which is used on my environment.

oomichi commented:
I also am using CentOS8.

champtar commented:
It's one commit or the other, no both at the same time

oomichi commented:
> 
> 
> It's one commit or the other, no both at the same time

I see, thanks.
I will try it again by each.

floryut commented:
Where are we on that ? 😃
